File: frontend/lib/client-labels.ts
import { API_BASE_URL } from "@/lib/api";
import { authHeaders } from "@/lib/auth";
import type { ClientLabel } from "@/lib/clients";

export type { ClientLabel } from "@/lib/clients";

export type ClientLabelList = {
  labels: ClientLabel[];
};

/**
 * Fetch every stored client label. Read-only; requires any authenticated role.
 */
export async function listClientLabels(): Promise<ClientLabelList> {
  const res = await fetch(`${API_BASE_URL}/clients/labels`, {
    headers: authHeaders(),
  });
  if (!res.ok) {
    throw new Error(`Labels request failed with status ${res.status}`);
  }
  return (await res.json()) as ClientLabelList;
}

/**
 * Create or update the label for a MAC (admin/write scope required). The MAC
 * may be in any common format; the backend normalizes it.
 */
export async function saveClientLabel(
  mac: string,
  label: string,
): Promise<ClientLabel> {
  const res = await fetch(
    `${API_BASE_URL}/clients/labels/${encodeURIComponent(mac)}`,
    {
      method: "PUT",
      headers: authHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify({ label }),
    },
  );
  if (!res.ok) {
    throw new Error(`Label save failed with status ${res.status}`);
  }
  return (await res.json()) as ClientLabel;
}

/**
 * Remove the label for a MAC (admin/write scope required).
 */
export async function deleteClientLabel(mac: string): Promise<void> {
  const res = await fetch(
    `${API_BASE_URL}/clients/labels/${encodeURIComponent(mac)}`,
    {
      method: "DELETE",
      headers: authHeaders(),
    },
  );
  if (!res.ok) {
    throw new Error(`Label delete failed with status ${res.status}`);
  }
}

File: backend/app/api/v1/client_labels.py
"""Client device-label endpoints.

Operator-assigned, persistent human-readable labels keyed by MAC address:

- ``GET    /clients/labels``     — list every label (any authenticated reader).
- ``PUT    /clients/labels/{mac}`` — create or update the label for a MAC.
- ``DELETE /clients/labels/{mac}`` — remove the label for a MAC.

Labels are application metadata only: they are stored in the local database,
never written into OpenWrt, and never change DHCP/ARP/WiFi configuration.
Writes require the admin/write scope (``devices.write``); reads follow the
existing authenticated read scope. MACs are normalized to the canonical
``aa:bb:cc:11:22:33`` form on every write, so equivalent representations map to
the same label.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from app.core.auth import AuthPrincipal, require_read, require_write
from app.db.client_label_store import ClientLabelStore, normalize_mac
from app.db.client_label_store import store as default_store

router = APIRouter(tags=["clients"])

#: Upper bound for a human-readable device label.
MAX_LABEL_LENGTH = 255


class LabelRequest(BaseModel):
    label: str = Field(min_length=1, max_length=MAX_LABEL_LENGTH)


def _store(request: Request) -> ClientLabelStore:
    """Resolve the label store (isolated stores can replace the default)."""
    return getattr(request.app.state, "client_label_store", None) or default_store


def _normalize_or_422(mac: str) -> str:
    normalized = normalize_mac(mac)
    if normalized is None:
        raise HTTPException(
            status_code=422,
            detail=f"Invalid MAC address: {mac!r}",
        )
    return normalized


@router.get("/clients/labels")
def list_labels(
    request: Request,
    principal: Annotated[AuthPrincipal, Depends(require_read)],
) -> dict:
    """Return every stored label, ordered by MAC address."""
    store = _store(request)
    labels = [store.to_public_dict(record) for record in store.list_all()]
    return {"labels": labels}


@router.put("/clients/labels/{mac}")
def put_label(
    request: Request,
    mac: str,
    body: LabelRequest,
    principal: Annotated[AuthPrincipal, Depends(require_write)],
) -> dict:
    """Create or update the label for a MAC (write/admin scope required)."""
    store = _store(request)
    normalized = _normalize_or_422(mac)
    label = body.label.strip()
    if not label:
        raise HTTPException(
            status_code=422,
            detail="Label must not be empty.",
        )
    record = store.upsert(normalized, label)
    return store.to_public_dict(record)


@router.delete("/clients/labels/{mac}")
def delete_label(
    request: Request,
    mac: str,
    principal: Annotated[AuthPrincipal, Depends(require_write)],
) -> dict:
    """Remove the label for a MAC (write/admin scope required)."""
    store = _store(request)
    normalized = _normalize_or_422(mac)
    if not store.delete(normalized):
        raise HTTPException(
            status_code=404,
            detail=f"No label stored for MAC {normalized}.",
        )
    return {"deleted": True, "mac_address": normalized}
